infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -283,6 +351,24 @@ private LogReplicationSession constructSession(String sourceClusterId, String si
                 .build();
     }
 
+    /**
+     * When a node acquires the lock, check if all the sessions present in the system tables are still valid.
+     *
+     * Since the metadata table is written to only by the leader node, we can run into the scenario where the leader node
+     * missed updating the system table because the leadership changed at that precise moment.
+     * If the new leader node had already received the topology change notification before becoming the leader, the stale
+     * sessions would not be recognized on "refresh".
+     * Therefore, iterate over the sessions in the system tables and remove the stale sessions if any.
+     */
+    public void removeStaleSessionOnLeadershipAcquire() {
+        try (TxnContext txn = metadataManager.getTxnContext()) {
+            Set<LogReplicationSession> sessionsInTable = txn.keySet(METADATA_TABLE_NAME);
+            sessionsInTable.stream().filter(sessionInTable -> !sessions.contains(sessionInTable))
+                    .forEach(staleSession -> metadataManager.removeSession(txn,staleSession));
+            txn.commit();
+        }
```

Comment:
Should we catch an exception and print a message?

Reply:
added retry to this logic